### ISSUE TITLE
Working directory independence

### ISSUE THREAD

**Issue by ruv** [SirWumpus/post4#86]
How to load a file, for example [test/assert.p4](https://github.com/SirWumpus/post4/blob/master/test/assert.p4), by full path, regardless of the working directory?

A desired pattern is:
```shell
post4 -i  /full/path/post4/test/assert.p4 -i ./my-program.fth
```

**Comment by SirWumpus** [comment]
The above should work just fine.  Alternatively put `assert.p4` into a `POST4_PATH` given directory.  `-i` works with `POST4_PATH` too.

Sideswipe: I should add `assert.p4` to the list of files installed into `/usr/local/lib/post4`, since its generally useful.

**Comment by ruv** [comment]
> The above should work just fine.

```transcript
post4 -i /work/forth.data/repo/github/post4/test/assert.p4

>> INCLUDE ../examples/ansiterm.p4
>>                                ^
2 thrown: ?
input: INCLUDE ../examples/ansiterm.p4
```
(a stack trace is not included)


> Alternatively put `assert.p4` into a `POST4_PATH` given directory. `-i` works with `POST4_PATH` too.

```transcript 
POST4_PATH=/work/forth.data/repo/github/post4/test/ post4 -i assert.p4

>> INCLUDE ../examples/ansiterm.p4
>>                                ^
2 thrown: ?
input: INCLUDE ../examples/ansiterm.p4
```

> I should add `assert.p4` to the list of files installed into `/usr/local/lib/post4`

Can it work without installing the files into special directories?

Actually, we discussed [possible solutions](https://github.com/SirWumpus/post4/discussions/74#discussioncomment-11335387) a little bit.

Note that in **most cases** you do not want to depend on the working directory when loading **a library** into the Forth system.  Typically, the path of the initial file of a library is either relative to the working directory or absolute, and all other paths are relative to the file being loaded.

**Comment by SirWumpus** [comment]
`INCLDUE` et al. should load relative the current directory or an absolute directory, per the standard.

`INCLUDE-PATH`, `REQUIRE-PATH`, et al. will load relative to POST4_PATH or an absolute path.

The test files have not yet been converted from using `INCLUDE` to `INCLUDE-PATH`.  `assert.p4` uses `INCLUDE ../examples/ansiterm.p4`.  Just need to update it to `INCLUDE-PATH` should solve.

**Comment by SirWumpus** [comment]
> Can it work without installing the files into special directories?

Doubt it.  If you install (as root) Post4 to a system you need to find the core file, post4.p4, some place without having to specify -c all the time.  Plus any other common .p4 libraries.

The JNI build have Java specific requirements concerning loading .class and .so files and the core files.  Putting them all together in one common place make sense.  Individual users can then define POST4_PATH to point at the root of their personal libraries.

Most package systems are no different, install system wide package some place, like `/usr/local/lib/post4/*`, user libraries in something like `$HOME/.post4` and probably per project libraries under `$HOME/path/to/project/root/.post4`.  Node's `npm`, Rust `cargo`, Erlang `rebar3`, Perl `cpan` etc. seem to behave similar.

**Comment by SirWumpus** [comment]
The `-i` option should allow from multiple possible choices with relative paths and absolute.  So in theory 
```
export POST4_PATH=".:/usr/local/lib/post4"
post4 -i pkg/printf/index.fth -i assert.p4 -i /opt/billybob/funnyforth/neatstuff.fth
```
should search
```
./pkg/printf/index.fth
/usr/local/lib/post4/pkg/printf/index.fth
./assert.p4
/usr/local/lib/post4/assert.p4
/opt/billybob/funnyforth/neatstuff.fth
```
Without `.` in the `POST4_PATH` you could skip the working directory.  When `POST4_PATH` is undefined, there is a [fall back](https://github.com/SirWumpus/post4/blob/810345face0e881d219209792b6fc80e81de2f01/src/post4.h#L90C1-L90C82) which include `.` directory.

I suppose I could add a `./configure` option to setup `P4_CORE_PATH` to make it easier to configure without fussing with environment variables.  Again a package maintainer would probably set the specifics for their OS (I made an educated guess at the most common for NetBSD, FreeBSD, and Linux).

**Comment by SirWumpus** [comment]
BTW `assert.p4` was originally intended to be part of the test suite only, so it just used `INCLUDE`.  Making it generally available just means I need to tidy it up a little.

**Comment by ruv** [comment]
> `INCLUDE-PATH`, `REQUIRE-PATH`, et al. will load relative to `POST4_PATH` or an absolute path.

This will not help loading libraries from **different packages**.

**Comment by ruv** [comment]
> Doubt it. If you install (as root) Post4 to a system you need to find the core file, `post4.p4`, some place without having to specify `-c` all the time. Plus any other common .p4 libraries.

I specify `-c` in the shell script once. 

BTW, can Post4 binary knows its own file location, from which it is started?

**Comment by SirWumpus** [comment]
> > `INCLUDE-PATH`, `REQUIRE-PATH`, et al. will load relative to `POST4_PATH` or an absolute path.
> 
> This will not help loading libraries from **different packages**.

Until I actually have some "packages" worth working with I won't know what other changes will be needed.  So far the 3 or 4 packages from `http://theforth.net/` I tried have dependencies, implementation specific words, and expected (but missing) optional word sets.  All the Forth devs seems to pride themselves on their own library setups, so I can't account for those.

**Comment by ruv** [comment]
> Until I actually have some "packages"

For example, the libraries/extensions provided with Post4 constitute a package, and within this package all relative paths are known.

If I load `/work/forth.data/repo/github/post4/test/assert.p4` by this full path, all required information **is known to the system** from this path.  Why should I specify `POST4_PATH` or anything else?

**Comment by ruv** [comment]
> > BTW, can Post4 binary knows its own file location, from which it is started?
> 
> It could, `getcwd(3)`,

I mean, if I start:
```shell
/work/forth.data/repo/github/post4/src/post4
```
can Post4 know that its binary can be found in the directory `/work/forth.data/repo/github/post4/src/`  ?

**Comment by SirWumpus** [comment]
> > > BTW, can Post4 binary knows its own file location, from which it is started?
> > 
> > 
> > It could, `getcwd(3)`,
> 
> I mean, if I start:
> 
> ```shell
> /work/forth.data/repo/github/post4/src/post4
> ```
> 
> can Post4 know that its binary can be found in the directory `/work/forth.data/repo/github/post4/src/` ?

Sure.  But what does that solve.  Plus when Post4 is installed, it would live in `/usr/local/bin`.

**Comment by ruv** [comment]
> > If I load `/work/forth.data/repo/github/post4/test/assert.p4` by this full path, all required information **is known to the system** from this path. Why should I specify `POST4_PATH` or anything else?
> 
> Because its a simple convenience, like `PATH`, `CDPATH`, `LD_LIBRARY_PATH`, `CLASSPATH` etc help find things (and save on typing lengthy absolute paths).

In some cases it is not convenience 🙃 

Well, could you please provide the words:
- `source-path ( -- sd.pathname )`
- `system-path ( -- sd.system-binary-pathname )`

And I will be able to implement everything else myself as a library.


-----

> I want an installed system to live under /usr/local as a general rule.

What if I want to test different versions of Post4 in parallel?


> Plus when Post4 is installed, it would live in /usr/local/bin

Reasonable. From this location of Post4 I could conclude that its additional files can be found in `/usr/local/lib/post4/`.

**Comment by ruv** [comment]
> `INCLDUE` et al. should load relative the current directory or an absolute directory, per the standard.

This is a common practice. But the standard only says: <q>Open the file specified by _c-addr u_</q>.

If the path is not absolute, a common practice is trying to open the file in different locations. Some systems try to open the file relative to the file being loaded, and/or to the system binary image location.

For example, see [`FIND-FULLNAME` in SP-Forth/4](https://github.com/rufig/spf4-utf8/blob/ab05ef34657d6f013ec0943b6c1513c2b4630c80/src/compiler/spf_translate.f#L283-L289).

**Comment by SirWumpus** [comment]
> Well, could you please provide the words:
> 
>     * `source-path ( -- sd.pathname )`

Source path?  Of?  The current include?

>     * `system-path ( -- sd.system-binary-pathname )`

The location of the post4 binary?  Without an absolute path that is harder to determine.

> And I will be able to implement everything else myself as a library.

Sounds like a cunning plan.

> > I want an installed system to live under /usr/local as a general rule.
> 
> What if I want to test different versions of Post4 in parallel?

Hmm.  Well that is a would be something for a clever shell script I suppose, similar to `nvm` or other tools for do the same for Python and Ruby, which is why I hate Python and Ruby development, monolith installs and switching between them- its a nightmare to manage and to use.

I suppose one of two choices:

```
/usr/local/post4/$POST4_MAJOR/
   bin/ lib/ doc/

or

/usr/local/post4-$POST4_MAJOR/
   bin/ lib/ doc/
```

But again that a different problem currently out of scope for Post4, more for a maintainer.  I'd be surprised if you need multiple versions.  The bulk of the words are in `post4.p4`, so using `-c` you should be able to select different word implementations.  Again it come down to directory tree.  Multiple versions built from source could configure P4_CORE_PATH based on the major.minor version, but personally following semver.org rules, only major releases should introduce breaking changes.

In theory, post4-1.0 interpreter should be stable; minor updates should just be fixing post4.p4.

But I haven't even reached that yet.  The JNI build might introduce more changes over time.


with some sort of Post4 selector in `/usr/local/bin/post4 [-M major]`



> 
> > Plus when Post4 is installed, it would live in /usr/local/bin
> 
> Reasonable. From this location of Post4 I could conclude that its additional files can be found in `/usr/local/lib/post4/`.

You could probably use `post4 -V` to extract details for scripting.  Similar constants could be made available in Forth, but I haven't thought much on that yet.  Maybe make a mod with `env` (really thinking I should rename it like the POSIX function `getenv`).

**Comment by SirWumpus** [comment]
> > `INCLDUE` et al. should load relative the current directory or an absolute directory, per the standard.
> 
> This is a common practice. But the standard only says: Open the file specified by _c-addr u_.
> 
> If the path is not absolute, a common practice is trying to open the file in different locations. Some systems try to open the file relative to the file being loaded, and/or to the system binary image location.
> 
> For example, see [`FIND-FULLNAME` in SP-Forth/4](https://github.com/rufig/spf4-utf8/blob/ab05ef34657d6f013ec0943b6c1513c2b4630c80/src/compiler/spf_translate.f#L283-L289).

You suggesting I merge `INCLUDE-PATH` et al. back into `INCLUDE` like I originally wanted?

**Comment by SirWumpus** [comment]
>  `source-path ( -- sd.path )`, to obtain an absolute path (in the general case a URI) of a file/device/document from which the input buffer is filled. It it is a file, then `open-file` should be able to open this file by this path.

Ok.  Figured it was something like that, just needed to be sure.

> A question is which path (if any) to return when a string is evaluated.

Well in theory (I use the word loosely, you seem to like mathematical theories), the URI would be `data: 123 345 + (do some forth stuff)`, where the URI is the string being evaluated string starting after the `data:`.  Since it  doesn't make sense and wastes space to repeat the whole thing for `source-path`, maybe shorten it to just `data:`.
 
> I like XML/HTML approach: https://www.w3.org/TR/xmlbase/#granularity

I hate XML and my eye glaze over; if you're ever in a lecture hall talking about XML snoring might ensue (look for me in the back).

> > > * `system-path ( -- sd.system-binary-pathname )`
> 
> > The location of the post4 binary?
> 
> Yes. So, if we pass this path to `open-file`, it should open this binary file. And if we pass this path to a shell or something like Win32 `CreateProcess`, we can lunch a new process.

I've already added this in one of the commits above.


> > Without an absolute path that is harder to determine.
> 
> What do you mean?

Well not all OSes support a `/proc` file system.  But NetBSD, Linux, and I recall FreeBSD do, which means I can find the path of the `exe`.

(Must add configuration test for this.)

> > Maybe make a mod with env (really thinking I should rename it like the POSIX function getenv).
> 
> Typically, the standard Forth word [`environment?`](https://forth-standard.org/standard/core/ENVIRONMENTq) returns values (strings) for the process environmental variables too.

Eww.  No.  `environment?` its ugly and ***deprecated***.  I prefer exposing the OS function for environment variables.  Its clear and well understood.

You might notice that every `environment?` constant is just a Forth constant in Post4 that avoid using `environment?` (which seems like an evil kludge from some arcane system).

**Comment by ruv** [comment]
> `environment?` its ugly 

Why?

> and **_deprecated_**.

Not really. Only the use of `environment?` to determine **whether a particular set of words is provided** by the system is deprecated. See also [3.2.7 Obsolescent Environmental Queries](https://forth-standard.org/standard/usage#subsection.3.2.7).

**Comment by ruv** [comment]
> I hate XML and my eye glaze over; 

Do you hate HTML as much as XML? 😉

> if you're ever in a lecture hall talking about XML snoring might ensue (look for me in the back).

😄

**Comment by SirWumpus** [comment]
>> environment? its ugly
> Why?


Because IMO it is a nonsensical way to determine any system constant, just make it a Forth constant.  Also for determining word sets, every word set has required words or constants that can be tested.

```
[defined] floating-stack [if] floating-stack . [then]

[defined] (local) [if] .... [then]
```
`environment?` never made any sense to me. Always seemed like a simpler more Forth like way is possible when it came to word sets and limits etc.  I was SO PLEASED when I found it was deprecated.  I know you said it is not completely, but for me its a dead word (like Latin).

**Comment by SirWumpus** [comment]
> > I hate XML and my eye glaze over;
> 
> Do you hate HTML as much as XML? 😉

I know HTML came from SGML (IBM as I recall).  And that HTML was a simplified version.  Then XML comes along to ruin everything IMO by over specify data (again) and being royal pain in the arse.

I did some work for a Italian finance data company and the trades were expressed in XML, badly.  And there could be huge number of trades per second, but real-time processing could not keep up, because parse XML using Java is slow.  The industry might have done better if they had standardised on JSON.

One colleague could be heard swearing in Italian across the room fairly regularly that summer.

Its a bit like ASN.1 specifications.  Ugly to read and parse, which is why they have a few different formats (XML being one).  I had to learn Erlang (which turned out to be fun) because I had to parse 3GPP CDRs and other types that were in PER format and Erlang was the ONLY open source language with an open source library capable of doing it (from PER).  I evaluated about 3 or 4 different commercial C based ASN.1 libraries.  It was easier and cheaper to learn a new language to handle ASN.1  Erlang has probably the most complete and comprehensive support.

BUT I digress.  XML is The Devil's excrement.

**Comment by ruv** [comment]
> Because IMO it is a nonsensical way to determine any system constant, just make it a Forth constant.

The idea is that these names shall be in a different name space. It does not prevent you to implement them as Forth words. And they are actually implemented as Forth words in a dedicated word list in some Forth systems — the word `environment?` checks this word list first, and if not found, checks the process environment variables. It's not much different from your `getenv` on that, and can be defined using it:

```forth
: environment? ( sd.name.standard-env -- i*x true |  sd.name.process-env -- sd.value true |  sd.any -- false )
  \ check the `env-wl` wordlist
  2dup env-wl find-name dup if nip nip name>interpret execute true exit then drop
  \ check the process environment variables
  getenv over 0<> if true exit then drop
;
```

**Comment by SirWumpus** [comment]
Oh.  Good 🙃  I can go take a nap now.

**Comment by ruv** [comment]
> > A question is which path (if any) to return when a string is evaluated.
> 
> Well in theory (I use the word loosely, you seem to like mathematical theories), the URI would be `data: 123 345 + (do some forth stuff)`, where the URI is the string being evaluated string starting after the `data:`. Since it doesn't make sense and wastes space to repeat the whole thing for `source-path`, maybe shorten it to just `data:`.
> 
> > I like XML/HTML approach: https://www.w3.org/TR/xmlbase/#granularity

I referred to XML/HTML, because they distinguish between the concepts of **base URI** and **document URI**.  Any relative URI in an element is resolved against the base URI of the element. The base URI of an element is calculated as follows (highest priority to lowest):
  - the explicitly specified base URI (if it is specified);
  - the base URI of the parent element (if it exists);
  - the base URI of the encapsulating entity (if it is specified) — this is the base URI of the document;
  - the URI used to retrieve the encapsulating entity (if it is available) — this is a document URI (after redirections, if any);
  - the base URI is defined by the context of the application.
See also [rfc3986 / Establishing a Base URI](https://www.rfc-editor.org/rfc/rfc3986.html#section-5.1).

-----

I think, the following phrases should produce the same result:
- `include ./foo.fth`
- `s" include ./foo.fth" evaluate`

This mean that a relative path must be resolved against the base URI, not against the document URI (which `source-path` returns).  And when starting evaluating a string, the base URI should be the same as before, despite `source-path` should return `( 0 0 )` or maybe some specific URI scheme.

A possible name for a word that returns the base path:
- `source-base-path ( -- sd.path )`

Other possible names (if we stick to the term "uri" instead of "path"):
- `source-uri`
- `source-base-uri`

Perhaps we should use another root/suffix instead of "base" to avoid possible confusing with the word `base`  (?)

What do you think?

**Comment by SirWumpus** [comment]
> > > A question is which path (if any) to return when a string is evaluated.
> > 
> > 
> > Well in theory (I use the word loosely, you seem to like mathematical theories), the URI would be `data: 123 345 + (do some forth stuff)`, where the URI is the string being evaluated string starting after the `data:`. Since it doesn't make sense and wastes space to repeat the whole thing for `source-path`, maybe shorten it to just `data:`.
> > > I like XML/HTML approach: https://www.w3.org/TR/xmlbase/#granularity
> 
> I referred to XML/HTML, because they distinguish between the concepts of **base URI** and **document URI**. Any relative URI in an element is resolved against the base URI of the element. The base URI of an element is calculated as follows (highest priority to lowest):

I understand the HTML \<BASE href=""> , used it some times.

> 
>     * the explicitly specified base URI (if it is specified);
> 
>     * the base URI of the parent element (if it exists);
> 
>     * the base URI of the encapsulating entity (if it is specified) — this is the base URI of the document;
> 
>     * the URI used to retrieve the encapsulating entity (if it is available) — this is a document URI (after redirections, if any);
> 
>     * the base URI is defined by the context of the application.
>       See also [rfc3986 / Establishing a Base URI](https://www.rfc-editor.org/rfc/rfc3986.html#section-5.1).

Always like RFCs.  Have to re-read.

> I think, the following phrases should produce the same result:
> 
>     * `include ./foo.fth`
> 
>     * `s" include ./foo.fth" evaluate`
> 
> 
> This mean that a relative path must be resolved against the base URI, not against the document URI (which `source-path` returns). And when starting evaluating a string, the base URI should be the same as before, despite `source-path` should return `( 0 0 )` or maybe some specific URI scheme.

What is the `source-path` of a block (`LOAD`)?

* string either `data:` or ( 0 0 )
* block maybe `block:NNN` or ( NNN 0 )
* file `/some/path/to/mandalay.fth` ( caddr u ) - I'd like to avoid including the use of `file:` for now (just more parsing to deal with).

> A possible name for a word that returns the base path:
> 
>     * `source-base-path ( -- sd.path )`
> 
> 
> Other possible names (if we stick to the term "uri" instead of "path"):

I only used `path` as it relates to POSIX file paths.

>     * `source-uri`
> 
>     * `source-base-uri`
> 
> 
> Perhaps we should use another root/suffix instead of "base" to avoid possible confusing with the word `base` (?)
> 
> What do you think?

By using `*-path` we can initially limit ourselves to the file system. Once you introduce `*-uri` variants then you get into supporting schemes `data:` `file:` `http:` `gopher:` etc. which brings in networking; I do want to go there eventually, but not yet, maybe after a v1.0.0.

To start lets confine ourselves to strings, blocks, and files.  Get that working sensibly.  

Also concerning networking, maybe we leave that the JNI build and interfacing with Java networking API instead of adding more internals to POSIX functions.  Not sure how I feel about that long term.

**Comment by ruv** [comment]
> What is the `source-path` of a block (LOAD)?

In the case of the input stream is the stdin, an evaluated string, a block, `source-path` should return a string that can be used in an error message, cannot be a file name or user-defined URI, and will be useful to the user. Probably, the [URI scheme `about`](https://en.wikipedia.org/wiki/About_URI_scheme) can be used for that, something like:
  - `about:source/stdin`
  - `about:source/string`
  - `about:source/block`

When a particular file is used to storing blocks, it may be indicated as well:
  - `about:source/block/path/to/block-file`

A block number can be indicated as an uri-fragment:
  - `about:source/block#132`
  - `about:source/block/full-path/to/block-file#132`

When interpreting a block or string, `source-base-path` should return the same value (a base URI) as before staring this interpretation.

-----

Some other words that can be useful in the source-* words family:

- `source-position ( -- u )`
Return the current position in the input buffer (that is `>in @`).
   - Rationale: `>in` is ugly and should be deprecated when a better alternative will be established.
- `source-following ( -- sd.string )`
Return the parse area, that is `source source-position /string`.
- `source-line ( -- u )`
Return the current line number (starting from 1), or 0 if a line number is not available.
   - for a file — the number for a line that is in the input buffer;
   - for a block — the line number within the input block to which the source-position belongs;
   - for a string — the line number within the string to which the source-position belongs if it is supported, otherwise 0;
      - a line terminator sequence belongs to the line it follows.
- `source-line-position ( -- u )`
Return the position in the line (starting from 1), or 0 if  a line number is not available.
  - this position can differ from `source-position` when the input source is a string or a block.
- `set-source-base-path ( sd.path -- )`
Set _sd.path_ to be used as the base path within the current input source.
   - this is necessary to implement some features like:
      - a markup to avoid repeating the same prefix in paths;
      - loading a file or a fragment of a file using `evaluate`.

-----

> To start lets confine ourselves to strings, blocks, and files. Get that working sensibly.

Agreed.

**Comment by ruv** [comment]
A few more ideas.

`about:system/lib/` can be resolved to the directory of the Forth system's libraries/extensions.
`about:system/wd/` can be resolved to the working directory.

So the following should work regardless of the working directory and the base path:
```forth
include about:system/lib/test/assert.p4
```

**Comment by SirWumpus** [comment]
> > What is the `source-path` of a block (LOAD)?
> 
> In the cases when the input stream is the stdin, an evaluated string, a block, `source-path` should return a string that can be used in an error message, cannot be a file name or user-defined URI, and will be useful to the user. Probably, the [URI scheme `about`](https://en.wikipedia.org/wiki/About_URI_scheme) can be used for that, something like:

Ok.  `about:` scheme makes sense, BUT I want to avoid any URI schemes on the initial go around stated above.

>     * `about:source/stdin`

Hmm.   Is there no `pipe:` scheme?  Because `stdin` could be a pipe instead of a terminal.

>     * `about:source/string`

While uniform, I still think if your going to use a scheme, the `data:` [RFC 2397](https://datatracker.ietf.org/doc/html/rfc2397) is the case here, eg. `data:text/plain,123 456 * . cr`.

Maybe make the concession to exclude the actual data, eg `data:text/plain`  

>     * `about:source/block`

Post4 evaluates blocks as 1024 byte strings, so again `data:` and MIME types could be applied, maybe
`data:text/block,123 456 + . cr ( lots more stuff upto 1024 )`.  Maybe a custom MIME type `data:x-block/999,123 456 + . cr ( lots more stuff upto 1024 )`

> A block number can be indicated as an uri-fragment:
> 
>     * `about:source/block#132`
> 
>     * `about:source/block/full-path/to/block-file#132`

Pretty.  Avoids including the actual data.

> When interpreting a block or string, `source-base-path` should return the same value (a base URI) as before staring this interpretation.

?? What would be the base path of a string or block? 

> Some other words that can be useful in the source-* words family:
> 
>     * `source-position ( -- u )`
>       Return the current position in the input buffer (starting from 0).

Is this not simply `>IN`?  While uniform, seems like duplication of functionality.

>       * That is `>in @`.

Hmmm.

>       * Rationale: `>in` is ugly and should be deprecated when a better alternative will be established.

Granted its ugly.  I think you could twist my rubber arm with that argument.  But you need to replace the inverse operation `123 >IN !` cleanly too.

>     * `set-source-position ( u -- )`
>       Set the position within the input buffer.

I really should read ahead before I spoke above.

Though I not sure I like separate words for get/put when you could just define `source-position ( -- addr )` then use `@` and `!`.

>  * That is `>in !`.
>  * If _u_ is greater than the number of primitive characters in the input buffer (i.e., the top value returned by `source`), the exception `-24` "invalid numeric argument" is thrown.
> 
>  * `source-following ( -- sd.string )`
>   Return the parse area, that is `source source-position /string`.

Hmm.  ok.  Are you cloning these ideas from elsewhere, like Gforth?   I've avoided looking as much as possible other implementations.

>     * `source-line ( -- u )`
>       Return the current line number (starting from 1), or 0 if a line number is not available or not supported.

Hmm.

>       * for a file — the number for a line that is in the input buffer;

Sure.

>       * for a block — the line number within the input block to which the source-position belongs;
>       * for a string — the line number within the string to which the source-position belongs;

Do not like this.  Post4 handles blocks as a string with fixed length lines.  I really want to avoid counting lines in blocks and strings.  Think its would be annoying to implement.

>  * a line terminator sequence belongs to the line it follows.
>  * Maybe, a better name is `source-row`  (?)

`source-index`  though I really dislike tracking lines outside of files.  Its easy with `REFILL` as line check point.   But strings and blocks, nope.

>  * `source-column ( -- u )`
>       Return the character number in the line (starting from 1) that corresponds to the position in the input buffer, or 0 if  not available or not supported.

BIG NO.  Post4 is not a text editor and I do not want to get into handling UTF8 just yet.  Having written text editors column tracking is a pain (tabs, multi-column/wide characters, multi-byte single column characters).  Ask again when I consider UTF8 support.

> * perhaps it should be a number of the _extended character_ if UTF (or Extended-Character word set) is supported.
> 
> * `set-source-base-path ( sd.path -- )`
>       Set _sd.path_ to be subsequently used as the base path within the current input source. If _sd.path_ is relative, it is resolved against the value returned by `source-base-path`.
>       
>  * this is necessary to implement some features like:
>         
>     * a markup to avoid repeating the same prefix in paths;
>     * loading a file or a fragment of a file using `evaluate`.

No convinced yet.

**Comment by ruv** [comment]
> >
> > * `set-source-position ( u -- )`
> >   Set the position within the input buffer.
> >
> 
> I really should read ahead before I spoke above.

😄  

> Though I not sure I like separate words for get/put when you could just define `source-position ( -- addr )` then use `@` and `!`.

The problem when a variable address is provided in API was discussed many times. The system cannot know the moment when new value is stored and cannot take the necessary actions. Among the standard words, this problem applies to `BASE`, `>IN`, `STATE`, `BLK`.

For example, `set-source-position` throws an exception when the parameter is incorrect.  `>IN !` cannot throw an exception.

**Comment by SirWumpus** [comment]
> A few more ideas.
> 
> `about:system/lib/` can be mapped to the directory of the Forth system's libraries/extensions. `about:system/wd/` can be mapped to the working directory.
> 
> So the following should work regardless of the working directory and the base path (base URI):
> 
> ```forth
> include about:system/lib/test/assert.p4
> ```

While it makes sense, this is why I want to avoid schemes on the first go around because of paring overhead of the scheme.  Plus its a little verbose.  What is wrong with:

```
about:lib/
about:cwd/
```

`cwd` well known abbrev taken from `getcwd()`.  Do you really need `system/` prefix?  Why not `post4/` (unless you're looking to standardise the behaviour and need a working example).

Hmm.  Too much to digest now.  Put this on the pile to look at again.

**Comment by ruv** [comment]
>  this is why I want to avoid schemes on the first go around because of paring overhead of the scheme.

Parsing overhead cost nothing comparing to translating a typical file.  Please, clarify.

**Comment by SirWumpus** [comment]
In LibSnert, `uri.c` (I always pronounce it like "Yuri"), I when to a lot of effort to deal with URI extraction from emails and parsing for `uri`, `BarricadeMX` and `milter-link`.  

For the curious, uri as CLI tool:
```shell
usage: uri [-ablLpqRsUv][-A delim][-d dbl,...][-i ip-bl,...][-m mail-bl,...]
           [-M domain-pat,...][-n ns-bl,...][-N ns-ip-bl,...][-u uri-bl,...]
           [-P port,...][-Q ns,...][-t sec][-T sec][arg ...][<input]

-a              check headers and body
-b              check body only (default)
-A delim        an alternative delimiter to replace the at-sign (@)
-d dbl,...      comma separate list of domain black lists
-i ip-bl,...    DNS suffix[/mask] list to apply. Without the /mask
                a suffix would be equivalent to suffix/0x00fffffe
-l              check HTTP links are valid & find origin server
-L              wait for all the replies from DNS list queries, need -v
-m mail-bl,...  DNS suffix[/mask] list to apply. Without the /mask
                a suffix would be equivalent to suffix/0x00fffffe
-M domain,...   list of domain glob-like patterns by which to limit
                checking against mail-bl; default *
-n ns-bl,...    DNS suffix[/mask] list to apply. Without the /mask
                a suffix would be equivalent to suffix/0x00fffffe
-N ns-ip-bl,... DNS suffix[/mask] list to apply. Without the /mask
                a suffix would be equivalent to suffix/0x00fffffe
-p              print each URI parsed
-P port,...     select only the URI corresponding to the comma
                separated list of port numbers to print and/or test
-q              check URL query part for embedded URLs
-Q ns,...       comma separated list of alternative name servers
-R              enable DNS round robin mode, default parallel mode
-s              check URI domain has valid SOA
-t sec          HTTP socket timeout in seconds, default 60
-T sec          DNS timeout in seconds, default 45
-u uri-bl,...   DNS suffix[/mask] list to apply. Without the /mask
                a suffix would be equivalent to suffix/0x00fffffe
-U              check sub-domains segments of URI domains
-v              verbose logging to system's user log

Each argument is either a filepath or string to be parsed and tested. If
the filepath "-" is given, then standard input is read. Also input can
be redirected.

Exit Codes
0               all URI tested are OK
1               one or more URI are blacklisted
64              usage error
70              internal error

LibSnert 1.75.58 Copyright 1996, 2019 by Anthony Howe. All rights reserved.
```

Always seems straight forward till someone starts waving URIs about until someone gets hit by URI and pokes out an eye.

Frankly I think Forth's string processing functions a little poor (compared to C which are average).

**Comment by ruv** [comment]
>  Plus its a little verbose. What is wrong with:
> ```
> about:lib/
> about:cwd/
> ```

Nothing wrong.  My idea is to put similar cases into subgroups (i.e., under the same path segment). If there are not many cases, subgroups are not needed.

Regarding `cwd` (which stands for "current working directory") and verbosity.  It seems, `wd` ("working directory") is enough, because there only one working directory, and "current" is excessive.

**Comment by ruv** [comment]
> I went to a lot of effort to deal with URI extraction from emails and parsing for uri, BarricadeMX and milter-link.

There is not need to parse all possible URI formats.  It is enough to parse only supported options and discard unsupported.

> Frankly I think Forth's string processing functions a little poor (compared to C which are average).

Have a look at my library [string-match.fth](https://github.com/ForthHub/fep-recognizer/blob/master/implementation/lib/string-match.fth).

An example or processing `about:` scheme and relative paths:
```forth
\ resolve-uri ( sd.uri.relative sd.uri.base -- sd.uri.resolved | sd.uri.absolute sd -- sd.uri.absolute )

: rewrite-path ( sd.path1 -- sd.path2 | 0 0 )
  dup 0= if exit then \ do not treat an empty path
  over c@ '/' = if exit then \ do not handle absolute path
  s" about:" match-head if
    s" lib/" match-head if  post4-library-path resolve-uri exit then
    s" wd/" match-head if exit then \ no need other rewrites for working directory
    2drop 0 0 exit \ no mapping
  then
  2dup s" :" contains if 2drop 0 0 exit then \ unsupported scheme
  source-base-path resolve-uri
;
\ the above is written straight from my head, not tested
```

**Comment by ruv** [comment]
>  though I really dislike tracking lines outside of files. Its easy with REFILL as line check point. But strings and blocks, nope.

> > * `source-column ( -- u )`
> >   Return the character number in the line (starting from 1) that corresponds to the position in the input buffer, or 0 if  not available or not supported.
> 
> BIG NO. Post4 is not a text editor and I do not want to get into handling UTF8 just yet.

`source-line` (or `source-row`) and `source-column` are only needed for friendly error reporting, and only when this information is requested.  So, the user will be able to open editor on a specified line number and character number in the line.  No need to calculate tabs, etc. Only number of characters (in supported charset).

If you do not support this info for blocks or strings, than these words just returns zero when a block or string is the input source.

I dislike the name `source-line` for that because "line" is associated with string of text. For example, `write-line`, `read-line`.
So, I choose `source-row`.  And then, `source-column` by association with row.

**Comment by SirWumpus** [comment]
> > though I really dislike tracking lines outside of files. Its easy with REFILL as line check point. But strings and blocks, nope.
> 
> > > * `source-column ( -- u )`
> > >   Return the character number in the line (starting from 1) that corresponds to the position in the input buffer, or 0 if  not available or not supported.
> > 
> > 
> > BIG NO. Post4 is not a text editor and I do not want to get into handling UTF8 just yet.
> 
> `source-line` (or `source-row`) and `source-column` are only needed for friendly error reporting, and only when this information is requested. So, the user will be able to open editor on a specified line number and character number in the line. No need to calculate tabs, etc. Only number of characters (in supported charset).
> 
> If you do not support this info for blocks or strings, than these words just returns zero when a block or string is the input source.
> 
> I dislike the name `source-line` for that because "line" is associated with string of text. For example, `write-line`, `read-line`. So, I choose `source-row`. And then, `source-column` by association with row.

Well `source-position` should be be enough if its possible to show enough context before and after.  Having the line number helps a lot, column less so.  I prefer the byte offset into the line.

Counting lines from stdin or a file would simple.

**Comment by ruv** [comment]
> > * `source-following ( -- sd.string )`
> >   Return the parse area, that is `source source-position /string`.
> 
> Hmm. ok. Are you cloning these ideas from elsewhere, like Gforth? I've avoided looking as much as possible other implementations.

No.  I have not seen this in any other Forth system yet. I only implemented that in libraries.  If these words prove useful and appear in several Forth systems, they could be proposed for standardization.
See also: [Portable line-oriented parsing](https://github.com/ForthHub/discussion/discussions/126) — I published implementation for some of them.

**Comment by ruv** [comment]
>  I prefer the byte offset into the line.

As I can see, text editors accepts not byte offset, but number of character in the line.

For example, see the command line option `--goto` in VSCode
https://code.visualstudio.com/docs/editor/command-line#_core-cli-options

**Comment by SirWumpus** [comment]
> > I prefer the byte offset into the line.
> 
> As I can see, text editors accepts not byte offset, but number of character in the line.
> 
> For example, see the command line option `--goto` in VSCode https://code.visualstudio.com/docs/editor/command-line#_core-cli-options

Eww.  VScode. In `vi(1)` it would be `<count>|`, where \<count> is number of characters.  Characters would be better than an offset, but unless you handle all input as wide-characters (which I don't at the moment), then it gets tricky after ASCII.

(I like ASCII, why can't the world adopt ASCII for everything, screw other multi-lingual symbols.)

When MKS internationalised all their software code, one of owners who studied the effort involved and said "it might just be cheaper to given them English<>YOURLANG dictionaries than update the code".

**Comment by ruv** [comment]
>>    When interpreting a block or string, source-base-path should return the same value (a base URI) as before staring this interpretation.

> ?? What would be the base path of a string or block?

```forth
t{ source-base-path s" source-base-path" evaluate compare 0= -> true }t
```
Inside an evaluated string, `source-base-path` returns a string that is equal to what it returns outside that evaluated string, no matter how deeply input sources are nested.

Ditto for blocks.

If not inside any file, and if its value is not specified using `set-source-base-path`, then `source-base-path` should return either empty string or the working directory  (do not sure what is better, but the effective result is the same).

**Comment by ruv** [comment]
> My idea is to put similar cases into subgroups (i.e., under the same path segment). If there are not many cases, subgroups are not needed.

Actually, there two subgroups at the moment:
1. mapping for some directories;
2. names for internal input sources (that are not files):
   - the user input device (e.g., `stdin`);
   - a block;
   - an evaluated string.

Names from the first and second group should be in different namespaces. An easiest way is to put them under different path segments. For example:
   - `about:path/`
      - `about:path/lib/`
      - `about:path/wd/`
   - `about:source/`
      - `about:source/stdin`
      - `about:source/string`
         - `about:source/string#9139860,19`
           - pattern: `about:source/string#{addres},{length}`
           - (just a possible format if you want to include address and length information for debugging purposes)
      - `about:source/block`
         - `about:source/block#132`
         - `about:source/block/full-path/to/block-file`
         - `about:source/block/full-path/to/block-file#232`

**Comment by SirWumpus** [comment]
> > My idea is to put similar cases into subgroups (i.e., under the same path segment). If there are not many cases, subgroups are not needed.
> 
> Actually, there two subgroups at the moment:
> 
>     1. mappings for system directories, that can be used in `included`;

✅

>     2. names that `source-path` returns for internal input sources (that are not files):
>        
>        * the user input device (e.g., `stdin`);
>        * a block;
>        * an evaluated string.

✅

> 
> 
> Names from the first and second subgroups should be in different namespaces. An easiest way is to 

✅

put them under different path segments. For example, place names from the first subgroup under 'path/', names from the second subgroup under 'input/':
> 
>     * `about:path/`
>       
>       * `about:path/lib/`
>       * `about:path/wd/`
> 
>     * `about:input/`

I like `input` over source, clearer in my mind;  otherwise I might have consider referring to output paths a `sinks`.  (And I hate doing dishes).

>       
>       * `about:input/stdin`
>       * `about:input/pipe`
>         
>         * `about:input/pipe#132`

What is the #132 portion?  Line number?  File descriptor?  Bytes read so far?

>       * `about:input/string`
>         
>         * `about:input/string#9139860,19`

✅ Should be stated address is in HEX.

>           * pattern: `about:input/string#{addres},{length}`
>           * (just a possible format if you want to include address and length information for debugging purposes)
>       * `about:input/block`
>         
>         * `about:input/block#132`
>         * `about:input/block/full-path/to/block-file`
>         * `about:input/block/full-path/to/block-file#232`

✅ (This would be Post4)

**Comment by ruv** [comment]
> What is the `#132` portion? Line number? File descriptor? Bytes read so far?

It is an optional id (or fileid) of the pipe that is used as the user input device, if this id is available and makes sense for debugging purposes (I do not sure).  Maybe, `about:input/stdin` is enough.

Probably, if this id is indicated, it should be a path segment rather than uri-fragment:  `about:input/pipe/132`

----

`source-path ( -- sd.path | 0 0 )` returns a path name (or URI) for the input source. It's a kind of **document URI**.

The idea is that if the input source does not have an external name (like file name), we use a name that starts with `about:input/`.

The string returned by `source-path` identifies entity that is **the input source**, not a line number or position in the input buffer.  So, it shall return the same value before and after `parse-name`.

-----

> (This would be Post4)

What do you mean?

**Comment by ruv** [comment]
> Would it not make more sense to use `/proc` file system? eg. `about:/proc/self/fd/0`

Good idea!  Probably, still under `input/`:  `about:input/proc/self/fd/0`

Alternatively, `source-path` may return `/dev/stdin` in this case — without `about:input/` prefix, because it's a file name in the real file system.  A possible problem is that when you include a file from interactive input, you do not want `/dev/stdin` be the base path. This should be handled.   When using the prefix `about:input/`, a general rule can be used that when some `about:` becomes the input source, it does not change the base path.


> I might have consider referring to output paths a `sinks`

Technically, it is also possible that `open-file` can accept some of such names too.

For example, at the moment there is no standard way to read raw stdin (only line by line using `accept`). In the same time, in POSIX you can open `/dev/stdin`, in Windows you can open `con`.  The name  `about:input/stdin` (or something alike) can be supported in a Forth system regardless OS.

**Comment by SirWumpus** [comment]
> * source-position ( -- u )
> * set-source-position ( u -- )

* `source-offset ( -- u )`
* `set-source-offset ( u -- )`

I been thinking that `source-offset` better reflects `>IN @` in that it is the offset within the input (source) buffer.  While `source-position` might be confused with absolute (file) position within a source stream.

**Comment by ruv** [comment]
> While `source-position` might be confused with absolute (file) position within a source stream.

I took "source-position" after "file-position".  Because it's a similar thing, but for the input buffer.

**Rationale:**  It is better to use the same term for similar things. The `source` returns the input buffer, then `source-position` returns the read position in the input buffer, similar to how `file-position` returns the read position in a file.

What do you think?


The position in the file that is the input source is not available for the program due to possible caching. Do we need the number of read chars (primitive characters) from the input source? It should take into account that the position in the input buffer can be rewound.

**Comment by SirWumpus** [comment]
Disagree.  Consider the wording of `>IN` talks of an offset.
> \>IN ( -- a-addr ) a-addr is the address of a cell containing the offset in characters from the start of the input buffer to the start of the parse area. 

`file-position` provides the absolute file location, which is why I question `source-position` as suitable name.  As you state getting an accurate position within a (pipe) stream is more or less informational since you cannot reposition back on a pipe.  Similarly, asking the user to retype from a given prior position might be asking a little much.

`source-offset` seems to make more sense as I see it, especially given the current description of `>IN`

**Comment by ruv** [comment]
> `file-position` provides the absolute file location

What do you mean by "absolute"?  It provides the _offset_ from the beginning of the file (that is the position of read from the file).
And `>in @` provides the _offset_ from the beginning of the input buffer (that is position of read/parse too).

```forth
s" /dev/null" r/o open-file throw file-position throw d.
```
— prints "0"  (the initial offset).


> `source-offset` seems to make more sense as I see it, especially given the current description of >IN

Well, let it be `source-offset`, if you still like it better :)

**Comment by SirWumpus** [comment]
> What do you mean by "absolute"? 

Well certainly not "relative".  I suppose I imply that with an absolute file position I can reposition at which the next read/write will take place.  

When dealing with buffers, I would think of an "offset" into the buffer.  Though semantically I suppose "position" could be used, but prefer to avoid that word.

> Well, let it be source-offset, if you still like it better :)

See!  I knew you would see it my way ... eventually.

**Comment by SirWumpus** [comment]
> source-following ( -- sd.string )
> Return the parse area, that is source source-position /string.

I already have this in Post4 as `source-remaining`.

**Comment by SirWumpus** [comment]
@ruv What is the expected life spans of `source-path` and `source-base-path`?

**Comment by ruv** [comment]
> What is the expected life spans of `source-path` and `source-base-path`?

### `source-path`

`source-path` returns identical character strings within the same input stream. The life span of the returned string should probably be until the end of translation of the input stream (withing which it has been returned). Another options is **to allow** the string to be returned in a transient region (and specify a shorter life span), if you have a rationale for that. Or, maybe you have a rationale to require that the returned string is the same (i.e., located at the same address).

### `source-base-path`

`source-base-path` may return different character strings within the same input stream. 

There two approaches to assign the base path:
- `set-source-base-path ( sd.path -- )`
   - it assigns the specified path to the base path;
   - the assigned path is used by the system until the next call to this word or until the end of the input stream (whichever comes first);
   - this word shall create a copy of the specified _sd.path_.
- `apply-base-path ( i*x xt sd.path -- j*x )`
   - it assigns the specified path to the base path, executes xt, and then restores the base path.

In the latter approach, we can require that the specified _sd.path_ shall be valid until `apply-base-path` returns. This simplifies memory management in the system and does not overly complicate programs. In addition, the program can be sure that the lifetime of the character string returned by `source-base-path`  lasts at least until the end of the containing definition invocation.

What do you think?

**Comment by ruv** [comment]
One question is: what `source-path` and `source-base-path` should return withing an invocation of `include-file`.

Possibly the most useful and expected behavior is following:
- `source-path` inside `include-file` returns either the actual path (if the system can obtain it from the fileid), or "about:input/file" otherwise;
- `source-base-path` inside `include-file` returns the same value as outside.
   - because it should not depend on limitations of a particular system.
This allows a program to use a pattern like:  ` ( fileid ) ['] include-file s" foo/bar.fth" apply-base-path`.

We can add a word like `include-file-path ( i*x fileid sd.path -- j*x )`, which uses the specified _sd.path_ for both `source-path` and `source-base-path`.  Concerning memory management, there are two options:
   - the character string _sd.path_ shall be valid until `include-file-path` returns;
   - the character string _sd.path_ may be transient, and the system shall make a copy (and free it by itself).

@SirWumpus, what do you think?

**Comment by ruv** [comment]
In all my use cases I need to set the base path for translating a lexical block (in a file, or in a string) or for `evaluate`, and then **restore** it . I cannot imagine a use case where you need to set the base path from some point in the input source until the end of the input source.

Thus, something like `apply-base-path` seems convenient for my use cases.


```forth
: apply-base-path ( i*x xt sd.path -- j*x )
  source-base-path 2>r  set-source-base-path
  catch
  2r> set-source-base-path   throw
;
```
Will this work?   Will the system be able to correctly manage resources in such a use case?

**Comment by SirWumpus** [comment]
> One question is: what `source-path` and `source-base-path` should return withing an invocation of `include-file`.
> 
> Possibly the most useful and expected behavior is following:
> 
>     * `source-path` inside `include-file` returns either the actual path (if the system can obtain it from the fileid), or "about:input/file" otherwise;

Currently (yesterday evening's push 4f6fbc2cc7) when `source-path` is known from `INCLUDE` or `INCLUDED`, then it is saved as the (absolute) path of the current source file, accessible from within the INCLUDED source.  See `INCLUDED` and `_include_file`.    

For `( fid ) INCLUDED-FILE` when the `fid` is opened by other means, then `source-path` is currently the empty string.  Not sure about the idea of `about:input/file`.  With the empty string its easy to detect and can be manipulated without testing (dirname, basename, etc.), whereas `about:input/file` requires some extra processing.


>     * `source-base-path` inside `include-file` returns the same value as outside.

✅       

>       * because it should not depend on limitations of a particular system.
>         This allows a program to use a pattern like:  ` ( fileid ) ['] include-file s" foo/bar.fth" apply-base-path`.

Not convinced that `apply-base-path` is necessary.

> We can add a word like `include-file-path ( i*x fileid sd.path -- j*x )`, which uses the specified _sd.path_ for both `source-path` and `source-base-path`. Concerning memory management, there are two options:

`include-file-path` sounds like what I implemented as `_include_file`.

However, `source-base-path` should remain untouched.
 
>     * the character string _sd.path_ shall be valid until `include-file-path` returns;

✅       

> 
>     * the character string _sd.path_ may be transient, and the system shall make a copy (and free it by itself).

✅       

> @SirWumpus, what do you think?

(Did I remember to turn off the stove after lunch?)

**Comment by SirWumpus** [comment]
I suppose I could redo `set-source-base-path` to same the string in allocated memory and hold the pointer, instead of static buffer, in which case your initial example would work.

**Comment by ruv** [comment]
> ```
> : apply-base-path
>    \ copy base path to allocated memory [ u, string ]
>    source-base-path path_max cell+ allocate dup >r dup cell+ swap set-filepath
>    set-source-base-path catch
>    \ restore base path, free the memory
>    r@ cell+ dup strlen set-source-base-path r> free drop
> ;
> ```
> 
> I suppose I could have implemented `set-filepath` better, but I think you'll get the idea.

Yes, I see.  This is the only possible approach if the string returned by `source-base-path`  has a short lifetime, or when its lifetime is impossible to predict in the general case.

**Comment by ruv** [comment]
> With the empty string its easy to detect and can be manipulated without testing (dirname, basename, etc.), whereas `about:input/file` requires some extra processing.

NB: this result is returned only if the system cannot obtain the path from a fileid.

In what use cases do you need to manipulate the result of `source-path`?
I only see its use in logging and error reporting, without a need to manipulate it. Moreover, `source-path` already returns "about:input/..." in some cases, so if you need manipulating, you need extra processing anyway.

**Comment by ruv** [comment]
> >   * because it should not depend on limitations of a particular system.
> >     This allows a program to use a pattern like:  ` ( fileid ) ['] include-file s" foo/bar.fth" apply-base-path`.

> Not convinced that `apply-base-path` is necessary.

Then, what equivalent code do you suggest instead of:
```forth
  ( fileid ) ['] include-file s" foo/bar.fth" apply-base-path
```
?

**Comment by SirWumpus** [comment]
> ```forth
>   ( fileid ) ['] include-file s" foo/bar.fth" apply-base-path
> ```

I'm confused (again).  `s" foo/bar.fth"` is not a base dir-path from which to look for relative files, but a relative file path.  

My understanding is `set-source-base-path` setup a directory path from which to search for related files like in a module/package (what name are you calling these Erlang they are modules, in most other languages packages).

And `apply-base-path` is something new only brought up today.  I'm still thinking about `source-path` `source-base-path` and how they would be used with `INCLUDE` or `INCLUDE-PATH`.

**Comment by SirWumpus** [comment]
> > With the empty string its easy to detect and can be manipulated without testing (dirname, basename, etc.), whereas `about:input/file` requires some extra processing.
> 
> NB: this result is returned only if the system cannot obtain the path from a fileid.

Yep.

> In what use cases do you need to manipulate the result of `source-path`? I only see its use in logging and error reporting, without a need to manipulate it. Moreover, `source-path` already returns "about:input/..." in some cases, so if you need manipulating, you need extra processing anyway.

No.  `source-path` currently doesn't return "about:input/file" for an unknown path, it currently returns "" - you want/asking it to return "about:input/file".  I see that as less useful.  And no I haven't thought of why a package might want to manipulate its own absolute file path, other than to navigate / find other directories or (resource) files relative to itself (html, libs, naughty pictures) (you know, the usual suspects).

**Comment by SirWumpus** [comment]
> Yes, I see. This is the only possible approach if the string returned by `source-base-path` has a short lifetime, or when its lifetime is impossible to predict in the general case.

I'm looking at implementing this in better way, but yes `source-base-path` can have a longer life than just the current INCLUDE.

**Comment by ruv** [comment]
> I'm confused (again). `s" foo/bar.fth"` is not a base dir-path from which to look for relative files, but a relative file path.

I am modeling the concept of a base path from XML/HTML (as I [wrote](https://github.com/SirWumpus/post4/issues/86#issuecomment-2518731868)). There a base URI can be a document URI (i.e., URI of a particular file). I assume the same for our base path.

> My understanding is `set-source-base-path` setup a directory path from which to search for related files like in a module/package

Allowing only a directory path is a possible option.  It simplifies an implementation/system a bit, and adds some bothering to a program.

> And `apply-base-path` is something new only brought up today. 

You asked about life spans, and I realized the difficulties of `set-source-base-path`, and that all use cases that I can imagine imply restoring the base path. A word that by itself restores the base path seems more convenient in use and simplifies memory management.

**Comment by ruv** [comment]
> > Moreover, `source-path` already returns "about:input/..." in some cases, so if you need manipulating, you need extra processing anyway.
> 
> No. `source-path` currently doesn't return "about:input/file" for an unknown path, it currently returns ""

For example, [it returns](https://github.com/SirWumpus/post4/blob/4f6fbc2cc7806bb9f67d9eaeb9652c9a604965c0/src/post4.c#L2595) "about:input/string" when the input source is an evaluated string. Thus, in the general case you need extra processing.

**Comment by SirWumpus** [comment]
> > I'm confused (again). `s" foo/bar.fth"` is not a base dir-path from which to look for relative files, but a relative file path.
> 
> I am modeling the concept of a base path from XML/HTML (as I [wrote](https://github.com/SirWumpus/post4/issues/86#issuecomment-2518731868)). There a base URI can be a document URI (i.e., URI of a particular file). I assume the same for our base path.

Well I'm going based on you early comment about `\<BASE href="">` which is a root dir. path.   I know NOTHING about XML and how it deals with this.

> > My understanding is `set-source-base-path` setup a directory path from which to search for related files like in a module/package
> 
> Allowing only a directory path is a possible option. It simplifies an implementation/system a bit, and adds some bothering to a program.

?? 

> > And `apply-base-path` is something new only brought up today.
> 
> You asked about life spans, and I realized the difficulties of `set-source-base-path`, and that all use cases that I can imagine imply restoring the base path. A word that by itself restores the base path seems more convenient in use and simplifies memory management.

Ah.  Ok.  In that context that makes sense now.  Sometimes I get a little fixated.

**Comment by SirWumpus** [comment]
> > > Moreover, `source-path` already returns "about:input/..." in some cases, so if you need manipulating, you need extra processing anyway.
> > 
> > 
> > No. `source-path` currently doesn't return "about:input/file" for an unknown path, it currently returns ""
> 
> For example, [it returns](https://github.com/SirWumpus/post4/blob/4f6fbc2cc7806bb9f67d9eaeb9652c9a604965c0/src/post4.c#L2595) "about:input/string" when the input source is an evaluated string. Thus, in the general case you need extra processing.

Well I'm not completely sold on that either.  I needed a place holder.  Though for a string, an empty would work too.

- - -
The other day I was trying to figure out how / when you could include from a string or block, since they have their own words.  Though...

```
s" data:text/plain, 123 456 +" INCLUDED
s" about:input/block/321" INCLUDED
```
could be the equivalent of
```
s" 123 456 +" EVALUATE
321 LOAD
```

Not sure it would buy you anything to support first case since their slightly more cumbersome versions (implementation wise), but possibly cleaner more modern replacements.

Do I make any sense?

**Comment by SirWumpus** [comment]
How about this implementation ...
```forth
\ X/Open PATH_MAX, see limits.h
1024 CONSTANT path_max

\ (S: <spaces>name -- )
: file-path CREATE 0 , path_max ALLOT DOES> @+ ;

\ (S: sd.path xt -- )
: set-file-path
	OVER 0 path_max 1- WITHIN 0= -11 AND THROW	
	EXECUTE DROP SWAP			
	2>R 2R@ strncpy	R> R> CELL- !
;

\ (S: sd.0 -- sd.1 )
: strndup DUP 1+ ALLOCATE THROW SWAP 2DUP 2>R strncpy 2R> ;

\ (S: -- sd.path )
file-path source-base-path

\ (S: -- sd.path )
: source-path _input_ptr @ in.path @ DUP strlen ; $02 _pp!

\ (S: sd.path -- )
: set-source-base-path ['] source-base-path set-file-path ; $20 _pp!

: apply-base-path ( i*x xt sd.path -- j*x )
	source-base-path strndup 2>r set-source-base-path
	catch
	2r> 2dup set-source-base-path drop free drop throw
;
```
Cleaner, clearer `apply-base-path` (than my earlier version).

**Comment by SirWumpus** [comment]
> You asked about life spans, and I realized the difficulties of set-source-base-path, and that all use cases that I can imagine imply restoring the base path. A word that by itself restores the base path seems more convenient in use and simplifies memory management.

Think about this comment.  

My take is that typically `set-source-base-path` is empty to start and that `_include_file` would set base path if necessary and clear it after if it was set by `_include_file`.  So there would be no need to nest `source-base-path`, unless your including package dependencies.

I still see `apply-base-path` as useful (now you explained why), but I would think its use infrequent.  Of course you might have other ideas with your packaging project.

**Comment by ruv** [comment]
> Though...
> 
> ```
> s" data:text/plain, 123 456 +" INCLUDED
> s" about:input/block/321" INCLUDED
> ```
> 
> could be the equivalent of
> 
> ```
> s" 123 456 +" EVALUATE
> 321 LOAD
> ```
> 
> Not sure it would buy you anything to support first case since their slightly more cumbersome versions (implementation wise), but possibly cleaner more modern replacements.
> 
> Do I make any sense?

It's interesting, because it's a uniform way. This capabilities can be supported for fun. But I don't expect anyone to compose special strings instead of using simpler options.
OTOH, if `charset` and `base64` are supported in the `data:` scheme, they can be useful in some rare cases (for example, nesting of translation without worrying about escaping characters).

**Comment by SirWumpus** [comment]
Yep [RFC 2397](https://datatracker.ietf.org/doc/html/rfc2397) allows for exactly that.

> It's interesting, because it's a uniform way. These capabilities can be supported for fun. 

When haven't we not had fun with Forth (asking for a friend).

**Comment by ruv** [comment]
I have created test cases, see:
https://github.com/ruv/fep-baseuri/blob/master/test/index.fth

A workflow that is expected to be operational:
```shell
git clone https://github.com/ruv/fep-baseuri.git && cd fep-baseuri
post4 -i about:path/lib/test/assert.p4 -i test/index.fth
```

**Comment by SirWumpus** [comment]
> ```shell
> git clone https://github.com/ruv/fep-baseuri.git && cd fep-baseuri
> post4 -i about:path/lib/test/assert.p4 -i test/index.fth
> ```

If you have installed Post4 then you should already be able to currently say:
```
post4 -i assert.p4 -i test/index.fth
```
And `assert.p4` will be found along one of the paths set for the default `POST4_PATH` as a library file.

**Comment by ruv** [comment]
> If you have installed Post4

You know, I don't want to install anything 🙃    I prefer to compiler and run.

How to set `POST4_PATH` so `post4 -i assert.p4 ` works from the git working tree?

-----

BTW currently at least one test fails, the last messages are:
```transcript
\ TEST: expected `source-path` inside `evaluate`

-9 thrown: invalid memory address
```

The following fails:
```forth
:noname source-path type ; s" execute" evaluate
```

**Comment by SirWumpus** [comment]
> > If you have installed Post4
> 
> You know, I don't want to install anything 🙃 I prefer to compiler and run.

Are you related to "Sheldon Cooper" by any chance?

> How to set `POST4_PATH` so `post4 -i assert.p4 ` works from the git working tree?

```
cd path/to/post4/src
cat <<EOF >post4.env
export POST4_ROOT=$(pwd | sed -e's,\(/post4\).*,\1,')
export POST4_PATH=".:$POST4_ROOT/examples:$POST4_ROOT/test"
EOF

. ./post4.env
post4 -i assert.p4
```

> BTW currently at least one test fails, the last messages are:
> 
> ```
> \ TEST: expected `source-path` inside `evaluate`
> 
> -9 thrown: invalid memory address
> ```
> 
> The following fails:
> 
> ```forth
> :noname source-path type ; s" execute" evaluate
> ```

Will look.  I've not tried your stuff yet.

**Comment by SirWumpus** [comment]
> > ```shell
> > export POST4_PATH=".:$POST4_ROOT/examples:$POST4_ROOT/test"
> > ```

> One problem with this approach is that you cannot have your own file `assert.p4` or `ansiterm.p4` in your working directory.

Yes you can, since the above path list contains `.` (current directory), so it will look in the current working directory before looking in the others.

> So I still think URIs like `about:lib/` make sense.

If you have `test/assert.p4` and are playing with changes in a local copy you could just tell it which one you want:
```shell
./post4 -i ./assert.p4 -i ../test/index.fth
# or
./post4 -i $(pwd)/assert.p4 -i ../test/index.fth
# or
./post4 -i ~/i/am/a/tea/pot/assert.p4 -i ../test/index.fth
```

Don't you ever play around shell scripts?

**Comment by ruv** [comment]
> Don't you ever play around shell scripts?

I tried the following:
```shell
cd /work/forth/2024/
echo '.( This file should not be used! ) cr' >ansiterm.p4
POST4_PATH=/work/forth.data/repo/github/post4/examples post4 /work/forth.data/repo/github/post4/test/assert.p4
```

Output:
```output
This file should not be used!

>> : test_pass 1 tests_passed +! ansi_green ." ." ansi_normal ;
>>                                         ^
-13 thrown: undefined word compiling test_pass
```

This is undesired behavior.  A library should not depend on accidental files in the working directory.

**Comment by SirWumpus** [comment]
> > Don't you ever play around shell scripts?
> 
> I tried the following:
> 
> ```shell
> cd /work/forth/2024/
> echo '.( This file should not be used! ) cr' >ansiterm.p4
> POST4_PATH=/work/forth.data/repo/github/post4/examples post4 /work/forth.data/repo/github/post4/test/assert.p4
> ```
> 
> Output:
> 
> ```
> This file should not be used!
> 
> >> : test_pass 1 tests_passed +! ansi_green ." ." ansi_normal ;
> >>                                         ^
> -13 thrown: undefined word compiling test_pass
> ```
> 
> This is undesired behavior. A library should not depend on accidental files in the working directory.

Hmm.  Ok.  The default `P4_CORE_PATH` includes the current directory.

https://github.com/SirWumpus/post4/blob/73243aa07459941128225e872711e6c16c1c24fa/src/post4.h#L97

I'll look to test it and change it.

**Comment by SirWumpus** [comment]
oops

**Comment by SirWumpus** [comment]
@ruv So this issue has gotten really long (with many tangents) and I'm starting to lose track of what's still on the table for consideration.  Have we achieved "working directory independence" yet?  Can related items be summarised in new issue(s)?

**Comment by ruv** [comment]
> Can related items be summarised in new issue(s)?

Yes, sure!

**Comment by ruv** [comment]
> Hmm. Ok. The default `P4_CORE_PATH` includes the current directory.

Do you mean that the default `P4_CORE_PATH` is used even if `POST4_PATH` is specified?

**Comment by ruv** [comment]
The [above test case](https://github.com/SirWumpus/post4/issues/86#issuecomment-2555332840) fails in aa45b5a

**Comment by SirWumpus** [comment]
> > Hmm. Ok. The default `P4_CORE_PATH` includes the current directory.
> 
> Do you mean that the default `P4_CORE_PATH` is used even if `POST4_PATH` is specified?

Yes.  You need a starting point.  Its similar to C's `/usr/include` being default for `#include <foo.h>` (thankfully C only defined one hard location), though `/usr/local/include` is popular for installed libs, but typically need `-I` to use it.

I need to know where to find `post4.p4` when no `$POST4_PATH` is set and I'm assuming the software is installed into `/usr/local` or similar package location (Linux is a pain since they just lump stuff together, not pleased by that).  I could prune `P4_CORE_PATH` to one location `/usr/local/lib/post4` and skip guessing what package maintainers might do, though allow for it to be a list still.

**Comment by SirWumpus** [comment]
> The [above test case](https://github.com/SirWumpus/post4/issues/86#issuecomment-2555332840) fails in [aa45b5a](https://github.com/SirWumpus/post4/commit/aa45b5a922f7b4ab194f54547e4db2fefe793c5f)

I fixed that yesterday.

**Comment by SirWumpus** [comment]
No further movement.  Closing.  Create a new issue if there are still parts outstanding.

**Comment by ruv** [comment]
> Some other words that can be useful in the source-* words family:

I have write them down, take a look in **test** branch:
https://github.com/ruv/fep-baseuri/blob/test/introspection/glossary.adoc

(will move them to the master after review)

**Comment by SirWumpus** [comment]
Plus `set-source-base-path`.

**Comment by ruv** [comment]
> Also I see no mention of `source-base-path`

Done: https://github.com/ruv/fep-baseuri/blob/master/readme.adoc

I've abandoned `set-source-base-path` in favor of `apply-base-path`, because the former complicates the process of freeing strings, without providing any benefit to programs.
